File: services/user-service/src/main/java/com/vnshop/userservice/infrastructure/web/ApiResponse.java
package com.vnshop.userservice.infrastructure.web;

import java.time.LocalDateTime;

public record ApiResponse<T>(
    boolean success,
    String message,
    T data,
    String errorCode,
    LocalDateTime timestamp
) {
    public static <T> ApiResponse<T> ok(T data) {
        return new ApiResponse<>(true, "Success", data, null, LocalDateTime.now());
    }

    public static <T> ApiResponse<T> ok(String message, T data) {
        return new ApiResponse<>(true, message, data, null, LocalDateTime.now());
    }

    public static <T> ApiResponse<T> error(String message, String errorCode) {
        return new ApiResponse<>(false, message, null, errorCode, LocalDateTime.now());
    }
}

File: services/cart-service/src/cart/infrastructure/api-response.ts
export class ApiResponse<T> {
  success: boolean;
  message: string;
  data: T | null;
  errorCode: string | null;
  timestamp: string;

  static ok<T>(data: T): ApiResponse<T>;
  static ok<T>(message: string, data: T): ApiResponse<T>;
  static ok<T>(messageOrData: string | T, maybeData?: T): ApiResponse<T> {
    const hasMessage =
      typeof messageOrData === 'string' && arguments.length === 2;

    return {
      success: true,
      message: hasMessage ? messageOrData : 'Success',
      data: hasMessage ? (maybeData as T) : (messageOrData as T),
      errorCode: null,
      timestamp: new Date().toISOString(),
    };
  }

  static error(message: string, errorCode: string): ApiResponse<null> {
    return {
      success: false,
      message,
      data: null,
      errorCode,
      timestamp: new Date().toISOString(),
    };
  }
}
